### PR TITLE
Warn when http sources are used in packages.config restore operations

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1519

Regression? Last working version:

## Description
Pretty straightforward. Contains the same warning as https://github.com/NuGet/NuGet.Client/pull/4552, but obviously with a code since it's restore.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by jeffkl** [comment]
Team Review: Add unit tests for the two other code paths where the warning is logged

**Comment by erdembayar** [comment]
> Team Review: Add unit tests for the two other code paths where the warning is logged

@NuGet/nuget-client  
I added tests for 2 other code paths.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 1]
```diff
@@ -224,6 +224,20 @@ private async Task PerformV2RestoreAsync(string packagesConfigFilePath, string i
...
+                IEnumerable<SourceRepository> sourceRepositories = packageRestoreContext.SourceRepositories;
+                IEnumerable<SourceRepository> enabledSources = (sourceRepositories != null && sourceRepositories.Any()) ? sourceRepositories :
+                    sourceRepositoryProvider.GetRepositories().Where(e => e.PackageSource.IsEnabled);
+
+                foreach (SourceRepository enabledSource in enabledSources)
+                {
+                    PackageSource source = enabledSource.PackageSource;
+                    if (source.IsHttp && !source.IsHttps)
+                    {
+                        projectContext.Log(RestoreLogMessage.CreateWarning(NuGetLogCode.NU1803,
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: We don't code the packages.config warnings. 
You don't need to log a restore message.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 2]
```diff
@@ -224,6 +224,20 @@ private async Task PerformV2RestoreAsync(string packagesConfigFilePath, string i
                     ClientPolicyContext = clientPolicyContext
                 };
 
+                IEnumerable<SourceRepository> sourceRepositories = packageRestoreContext.SourceRepositories;
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: Don't you have the sources already in line 170ish? 

You are using the context, but the context is actually created just above in the same method.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 3]
```diff
@@ -224,6 +224,20 @@ private async Task PerformV2RestoreAsync(string packagesConfigFilePath, string i
...
+                {
+                    PackageSource source = enabledSource.PackageSource;
+                    if (source.IsHttp && !source.IsHttps)
+                    {
+                        projectContext.Log(RestoreLogMessage.CreateWarning(NuGetLogCode.NU1803,
+                            string.Format(CultureInfo.CurrentCulture, LocalizedResourceManager.GetString("Warning_HttpServerUsage"), "restore", source.Source)));
+                    }
+                }
+
                 var result = await PackageRestoreManager.RestoreMissingPackagesAsync(
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: Is it possible to warn in PackageRestoreManager instead? That way you capture all scenarios.

**Review by nkolev92 — COMMENTED** [review]
You're missing packages.config restore in Visual Studio.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: Call to `PackageRestoreManager` happen later for each packagereference, so it'll duplicate the warning.
That is why previously added `packageSourceMapping` logic just above here line 220ish too.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: Not sure I follow. In this codepath, it seems like the only place where this is called.

**Comment by erdembayar** [comment]
> You're missing packages.config restore in Visual Studio.

I just tested PR restore in Visual Studio and I don't see any warning in "Package Manager console".
It seems restore logging in Visual Studio has a bug.
When I'm debugging packages.config restore in VS, I can see it's logging something but nothing in "Package Manager console".
image
I can file separate bug for this. Does it sound fair?

**Comment by nkolev92** [comment]
> I just tested PR restore in Visual Studio and I don't see any warning in "Package Manager console".

I assume you mean PC restore right? 

> When I'm debugging packages.config restore in VS, I can see it's logging something but nothing in "Package Manager console".

Are you looking in PMC?
We have an output window titled `Package  Manager`, check there.

**Comment by erdembayar** [comment]
> > I just tested PR restore in Visual Studio and I don't see any warning in "Package Manager console".
> 
> I assume you mean PC restore right?
> 
> > When I'm debugging packages.config restore in VS, I can see it's logging something but nothing in "Package Manager console".
> 
> Are you looking in PMC? We have an output window titled `Package Manager`, check there.

Sorry for naming confusion let me clear it for you. 

**1. PMUI scenario**
I tested both PC and PR with `PMUI install` scenario, both cases I don't see any warning in Output >> PackageManager, I suspect there is a logging bug here. Would you like me file bug for it?
PR:
image

PC:
image

**2. PMC scenario**
Then I tested both PC and PR in `PMC` window

PR:  I can see asset file include text2xml entry, that means restore did happen.
image

PC:
image

By the way I did clean up cache between repro cycles and I used below 2 sources for tests:
image

**Comment by nkolev92** [comment]
Installation in packages.config is different from installation in PackageReference.
We discussed shortly in one of the teams channels, it's all restore in PackageReference. In packages.config, they are different gestures. 

We'll have a different issue for installation scenarios.

**Comment by nkolev92** [comment]
I originally thought your branch didn't have my changes, but it seems like now does. How are you testing that the PR scenarios?

I'll do some tests on my end, as that's something I expect to work.

**Comment by erdembayar** [comment]
> I originally thought your branch didn't have my changes, but it seems like now does. How are you testing that the PR scenarios?
> 
> I'll do some tests on my end, as that's something I expect to work.

I just tested in VS PMUI and VS PMC with below 2 feeds.
```
    <add key="nuget" value="http://api.nuget.org/v3/index.json" />
    <add key="myinsecure" value="http://nuget.org/api/v2" />
```

**Comment by nkolev92** [comment]
My question is more about what version of VS you are testing. Are you using experimental instance etc.

**Comment by erdembayar** [comment]
> My question is more about what version of VS you are testing. Are you using experimental instance etc.

Yes, it's experimental instance for VS 17.2 Preview 4.
image

**Comment by nkolev92** [comment]
@erdembayar 

Try using a legacy package reference project. 

For SDK based projects, the warnings are added in by the SDK and given that the SDK doesn't know about NU1803 it can't parse it. 

Root cause explained: https://github.com/NuGet/Home/issues/7126

**Comment by erdembayar** [comment]
> @erdembayar
> 
> Try using a legacy package reference project.
> 
> For SDK based projects, the warnings are added in by the SDK and given that the SDK doesn't know about NU1803 it can't parse it.
> 
> Root cause explained: [NuGet/Home#7126](https://github.com/NuGet/Home/issues/7126)

You're right it works for legacy PR warnings are added.
image
image

Does above apply for package PMUI install of packages.config projects too?

**Comment by nkolev92** [comment]
> Does above apply for package PMUI install of packages.config projects too

What does above mean?

**Comment by erdembayar** [comment]
> > Does above apply for package PMUI install of packages.config projects too
> 
> What does above mean?

Install package to PC project via PMUI, I don't see any warning, but it's not sdk based so just wandering why:
image

**Comment by erdembayar** [comment]
> > > Does above apply for package PMUI install of packages.config projects too
> > 
> > 
> > What does above mean?
> 
> Install package to PC project via PMUI, I don't see any warning, but it's not sdk based so just wandering why: ![image]

Never mind, I was using wrong logger, I changed it. Now both PMC and PMUI warnings are working.
image

image

It's ready to review.

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
It's ready to review.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs [hunk 4]
```diff
@@ -255,5 +257,51 @@ public static CommandRunnerResult RunInstall(string nugetExe, SimpleTestPathCont
...
+            var packageA = new SimpleTestPackageContext("a", "1.0.0");
+            await SimpleTestPackageUtility.CreateFolderFeedV3Async(pathContext.PackageSource, packageA);
+            var packageAPath = Path.Combine(pathContext.PackageSource, "a", "1.0.0", "a.1.0.0.nupkg");
+
+            pathContext.Settings.AddSource("http-feed", "http://api.source/api/v2");
+            pathContext.Settings.AddSource("https-feed", "https://api.source/index.json");
+
+            var net461 = NuGetFramework.Parse("net461");
+
+            var projectA = SimpleTestProjectContext.CreateLegacyPackageReference(
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs]: Shouldn't use a LegacyPackageReference project, but just a framework project here.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs [hunk 5]
```diff
@@ -1173,6 +1173,43 @@ public async Task Restore_PackageSourceMapping_Fails()
...
+            // Set up solution, project, and packages
+            var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+            var packageA = new SimpleTestPackageContext("a", "1.0.0");
+            await SimpleTestPackageUtility.CreateFolderFeedV3Async(pathContext.PackageSource, packageA);
+            pathContext.Settings.AddSource("http-feed", "http://api.source/index.json");
+            pathContext.Settings.AddSource("https-feed", "https://api.source/index.json");
+
+            var net461 = NuGetFramework.Parse("net461");
+
+            var projectA = SimpleTestProjectContext.CreateLegacyPackageReference(
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs]: Same here, please don't use a legacy package reference project.

File: test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs [hunk 6]
```diff
@@ -969,5 +969,59 @@ public async Task MsbuildRestore_WithCPPCliVcxproj_WithAssetTargetFallback_Succe
...
+                    ProjectStyle.PackagesConfig,
+                    pathContext.SolutionRoot);
+                projectA.Frameworks.Add(new SimpleTestProjectFrameworkContext(net461));
+
+                var packageX = new SimpleTestPackageContext()
+                {
+                    Id = "x",
+                    Version = "1.0.0"
+                };
+                packageX.Files.Clear();
```
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs]: you don't need the clear.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs [hunk 7]
```diff
@@ -1173,6 +1173,43 @@ public async Task Restore_PackageSourceMapping_Fails()
...
+
+            solution.Projects.Add(projectA);
+            solution.Create(pathContext.SolutionRoot);
+
+            // Act
+            var result = RunRestore(pathContext, _successExitCode);
+
+            // Assert
+            result.Success.Should().BeTrue();
+            Assert.Contains("You are running the 'restore' operation with an 'http' source, 'http://api.source/index.json'. Support for 'http' sources will be removed in a future version.", result.Output);
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:1211]: Maybe add the validaiton of `                Assert.Contains("Added package 'x.1.0.0' to folder", result.AllOutput);` in here as well?

In particular, you want to make sure a PC restore is running.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs [hunk 8]
```diff
@@ -255,5 +257,51 @@ public static CommandRunnerResult RunInstall(string nugetExe, SimpleTestPathCont
...
+                Path.Combine(Path.GetDirectoryName(projectA.ProjectPath), "packages.config"),
+                "-OutputDirectory",
+                Path.Combine(pathContext.WorkingDirectory, "packages")
+            };
+
+            // Act
+            var result = RunInstall(_nugetExePath, pathContext, expectedExitCode: 0, additionalArgs: args);
+
+            // Assert
+            result.Success.Should().BeTrue();
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs:304]: Maybe add the validaiton of `                Assert.Contains("Added package 'x.1.0.0' to folder", result.AllOutput);` in here as well?

In particular, you want to make sure a PC restore is running.

File: src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs [hunk 9]
```diff
@@ -377,6 +378,22 @@ private NuGetPackageManager GetNuGetPackageManager(string solutionDirectory)
 
             packageRestoreContext.Token.ThrowIfCancellationRequested();
 
+            IEnumerable<SourceRepository> enabledSources =
+                packageRestoreContext
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs]: packageRestoreContext has the SourceRepositories. Can we use those?

**Comment by nkolev92** [comment]
When you say you `PMC and PMUI warnings` work, can you clarify what commands you are running there?

**Review by nkolev92 — COMMENTED** [review]
Looks pretty good overall.

File: src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs (hunk 9 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs]: I think we can use `packageRestoreContext.SourceRepositories` because the collection is populated with sources that are enabled.

https://github.com/NuGet/NuGet.Client/blob/aca94657c6cf558c69bb3b5228856581c8c5b210/src/NuGet.Core/NuGet.Protocol/CachingSourceProvider.cs#L33-L36

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs [hunk 10]
```diff
@@ -36,7 +36,7 @@ public class NuGetPackageManager
     {
         private IReadOnlyList<SourceRepository> _globalPackageFolderRepositories;
 
-        private ISourceRepositoryProvider SourceRepositoryProvider { get; }
+        public ISourceRepositoryProvider SourceRepositoryProvider { get; }
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs]: I guess we can revert this change or make it `internal` if we use `packageRestoreContext.SourceRepositories` as per https://github.com/NuGet/NuGet.Client/pull/4583/files#r859185269

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs (hunk 4 above)
Comment (kartheekp-ms) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs]: I found an example in test code on how to create a framework project. Hope it helps.
``` cs
var projectA = new SimpleTestProjectContext(
                    "a",
                    ProjectStyle.PackagesConfig,
                    pathContext.SolutionRoot);
                projectA.Frameworks.Add(new SimpleTestProjectFrameworkContext(net461));
```

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs (hunk 10 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs]: Reverted.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs (hunk 8 above)
Comment (erdembayar) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/InstallCommandTests.cs:304]: Sure.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs (hunk 7 above)
Comment (erdembayar) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:1211]: Sure.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs (hunk 5 above)
Comment (erdembayar) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs]: Sure.

**Comment by erdembayar** [comment]
@nkolev92 @kartheekp-ms 
I addressed PR comments.

**Review by nkolev92 — DISMISSED** [review]

File: src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs (hunk 9 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs]: I had to add back this change, because some code paths `sourceRepositories` is set explicitly null so we can't use `packageRestoreContext.SourceRepositories` that case.

https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs#L312

**Review by jeffkl — DISMISSED** [review]

File: src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs (hunk 9 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs]: I'd really love it if we can avoid adding *any* public methods in NuGetPackageManager. 

I tried to follow RestorePackageAsync into NuGetPackageManager and it uses the ISourceRepositoryProvider to get the providers if they are not set on the context. 

Maybe we can just set them?

File: src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs [hunk 11]
```diff
@@ -377,6 +378,15 @@ private NuGetPackageManager GetNuGetPackageManager(string solutionDirectory)
 
             packageRestoreContext.Token.ThrowIfCancellationRequested();
 
+            foreach (SourceRepository enabledSource in packageRestoreContext.SourceRepositories)
+            {
+                PackageSource source = enabledSource.PackageSource;
+                if (source.IsHttp && !source.IsHttps)
+                {
+                    packageRestoreContext.Logger.Log(LogLevel.Warning, string.Format(CultureInfo.CurrentCulture, Strings.Warning_HttpServerUsage, "restore", source.Source));
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs:386]: I think we need to add log code to the warning. Similar to PR restore logic. https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs#L223-L224

Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs:386]: log codes only work in PackageReference.
For consistency we try not to add the log codes in packages.config as none of the resolver warnings/errors have any codes.

Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs:386]: > For consistency we try not to add the log codes in packages.config as none of the resolver warnings/errors have any codes.

I don't understand this. With a log code, the customer has something to search for (or in VS, something to click in the Error List window), and then there can be a docs page where we can provide more detailed instructions on how to resolve the error.

Why is consistency with not using log codes matter more than this?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs:386]: You can't suppress warnings or elevate warnings in packages.config. Log codes usually suggest that you can do that. 
Also warnings/errors in packages.config and PackageReference are different because pc restore rarely puts any log messages there. 

I think the message itself should be actionable enough.

**Review by nkolev92 — DISMISSED** [review]
